docs: Overhaul user data formats documentation

TheRealFalcon opened the pull request:
<!--
Thank you for submitting a PR to cloud-init!

To ease the process of reviewing your PR, do make sure to complete the following checklist **before** submitting a pull request.

- [ ] I have signed the CLA: https://ubuntu.com/legal/contributors
- [ ] I have added my Github username to ``tools/.github-cla-signers``
- [ ] I have included a comprehensive commit message using the guide below
- [ ] I have added unit tests to cover the new behavior under ``tests/unittests/``
  - Test files should map to source files i.e. a source file ``cloudinit/example.py`` should be tested by ``tests/unittests/test_example.py``
  - Run unit tests with ``tox -e py3``
- [ ] I have kept the change small, avoiding unnecessary whitespace or non-functional changes.
- [ ] I have added a reference to issues that this PR relates to in the PR message (Refs GH-1234, Fixes GH-1234)
- [ ] I have updated the documentation with the changed behavior.
  - If the change doesn't change the user interface and is trivial, this step may be skipped.
  - Cloud-config documentation is generated from the jsonschema.
  - Generate docs with ``tox -e docs``.
-->


## Proposed Commit Message
<!-- Include a proposed commit message because PRs are squash merged
by default.

See https://www.conventionalcommits.org/en/v1.0.0/#specification
for our commit message convention.

If the change is related to a particular cloud or particular distro,
please include the "optional scope" in the summary line. E.g.,
feat(ec2): Add support for foo to the baz

Types used by this project:
feat, fix, docs, ci, test, refactor, chore
-->
```
docs: Overhaul user data formats documentation

Fixes GH-4739
```

## Additional Context
I kept (and updated) the part handler piece as I think it should stay here until https://github.com/canonical/cloud-init/issues/4649 is implemented. 

I started documenting cloud-config-jsonp, but decided against it and filed https://github.com/canonical/cloud-init/issues/5549 instead. I don't think we should be using it or recommending it to anybody. I also updated the vendor data page accordingly to remove the suggestion there.

There's unfortunately some significant overlap with https://github.com/canonical/cloud-init/pull/5546 . I stole some of the language there because a full-on cherry-pick would have been too hard to merge. If the commit there can be simplified, I'm happy to cherry-pick it and rebase mine on top to give Alberto some credit for his additions.

I'm not married to the format I chose, so if a reviewer has any better ideas, feel free to suggest.

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->


## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

s-makin commented on doc/rtd/explanation/format.rst:
```diff
@@ -3,18 +3,65 @@
-launch configure the instance. User data can be one of the following types.
+User data is configuration data provided by a user of a cloud platform to an
+instance at launch. User data can be passed to cloud-init in any of many
+formats documented here. Each format will document a required header or
+content-type that must be present in the user data to be recognized as that
+format.
```
"Each format will document a required header or content-type that must be present in the user data to be recognized as that format."
I found this a bit confusing - particularly "each format will document". Is the format doing the documenting? Or is it that each format requires a specific header or content type?

TheRealFalcon replied:
I updated that sentence. Let me know if that's any better.

s-makin replied:
It's a lot clearer to me :) thanks!

s-makin commented on doc/rtd/explanation/format.rst:
```diff
@@ -3,18 +3,65 @@
+| **Header:** #cloud-config
+| **Content-Type:** text/cloud-config
+
+Example
+-------
```
I don't know that we need this or the explanation header to be subheaders. For consistency with the header and content type, would it make sense to set these to **Example** and **Explanation**? The sections don't feel long enough to justify having subsections in them.

TheRealFalcon replied:
I think I prefer subheadings rather than bold, but I pushed a bold commit for comparison.

s-makin replied:
It's nothing I'd be willing to hold up the doc over. If you're happier with the subheadings, we'll leave them as they are.

s-makin reviewed (commented):
Overall I really love this change. I think it feels more standardised and definitely a lot more straightforward, and the consistent format for each section in the format page makes it easier to parse the information. Just have one small suggestion, but otherwise lgtm :)

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -141,66 +225,128 @@ Create user data containing 3 shell scripts:
+Explanation
+-----------
+
+A cloud-config-archive is a way to specify more than one type of data
```
Are there any benefits over a MIME multi-part archive? Why would anyone want to use this over MIME?

TheRealFalcon replied:
Why would anyone NOT want to use this over MIME? Have you tried building one yourself? Do you like calling helper commands to build your user data? :wink: 

I can add a "for those that prefer to use YAML" to the end of this section. Are you wanting more of an explanation than that in the docs?

holmanb replied:
Yeah, I get it :)

>  It can be seen as an alternative to building a MIME multi-part archive manually.

We present two alternatives, without any recommended / preferred choice or cost / benefit analysis to the end user. If they've tried MIME before, then they'll know, but if not then they have to learn the hard way. Maybe we could save them the trouble?

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -141,66 +225,128 @@ Create user data containing 3 shell scripts:
+| **Header:** #cloud-config-archive
+| **Content-Type:** text/cloud-config-archive
+
+Example
```
Nice example!

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -141,66 +225,128 @@ Create user data containing 3 shell scripts:
+
+    #cloud-config-archive
+    - type: "text/cloud-boothook"
+    content: |
```
should this be indented further?

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -141,66 +225,128 @@ Create user data containing 3 shell scripts:
+        #!/bin/sh
+        echo "this is from a boothook." > /var/tmp/boothook.txt
+    - type: "text/cloud-config"
+    content: |
```
should this be indented further?

holmanb commented on doc/examples/part-handler.txt:
```diff
@@ -1,22 +1,58 @@
+def handle_part(data: Cloud, ctype: str, filename: str, payload: Any):
+    """Handle a part with the given mime-type.
+
+    This function will get called mulltiple times. The first time is
```
sp: multiple

holmanb commented on doc/examples/part-handler.txt:
```diff
@@ -1,22 +1,58 @@
 #part-handler
 
+"""This is a trivial example part-handler that creates a file with the path
```
Much needed context throughout this file, thanks for adding it!

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -141,66 +225,128 @@ Create user data containing 3 shell scripts:
 
-Begins with: ``#include`` or ``Content-Type: text/x-include-url``  when using
-a MIME archive.
+* ``type``: The content type of the MIME part
```
This format doesn't use a MIME part, does it? Maybe something like this instead?

> type: The Content-Type identifier for the type of user data in content
> content: The user data configuration

TheRealFalcon replied:
It translates fairly directly to a mime part that gets generated from the config. I can pull the MIME details out of this section, but IMO it'll make some of the other fields harder to understand.

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -141,66 +225,128 @@ Create user data containing 3 shell scripts:
+config and meta config. If an error occurs reading a file the remaining files
+will not be read.
+
+Gzip compressed content
```
Is this format supported on all platforms? I seem to recall there being some platform-specific oddities about this but don't remember specifics.

TheRealFalcon replied:
As far as I know it should. If your cloud mangles your user data (like LXD does), it won't work, but that's the only limitation I'm aware of.

holmanb reviewed (changes requested):
I really like this work! The organization and standardization is much improved. It is much easier to read and adds lots of context.

I left some suggestions and questions inline.

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -141,85 +216,150 @@ Create user data containing 3 shell scripts:
-================
 
-This content is an :file:`include` file.
+Cloud config archive
```
Aside: is cloud config archive nesting supported? I've not checked the code, but I'm curious whether I could stuff have:

- cloud config archive -> cloud config archive -> cloud config archive
- cloud config archive -> mime -> cloud config archive
- mime -> cloud config archive -> mime

and where/when things might start to break with nesting

TheRealFalcon replied:
I'll leave that as an exercise for you to figure out :joy:

TheRealFalcon commented on doc/rtd/explanation/format.rst:
```diff
@@ -3,18 +3,65 @@
+Configuration types
+===================
+
+User data can be categorized into **base config**
```
What do you think about not trying to define them at all. This section could be updated to say something like

User data formats can be split into those that  directly configure the instance and those that serve as a container, template, or means to obtain or modify another configuration.

Formats that directly configure the instance:
...

Formats that deal with other user data formats:
...

holmanb replied:
I can see it being useful to have a name for these than "formats that directly configure the instance", but I'm fine with what you've suggested if that's what you would prefer.

TheRealFalcon commented on doc/rtd/explanation/format.rst:
```diff
@@ -24,64 +71,101 @@ These things may include:
 
-This list of rules is applied to each part of this multi-part file.
+| **Header:** Content-Type: multipart/mixed;
+| **Content-Type:** multipart/mixed
```
Heh. The header is the content-type declaration...so...it's a bit weird.

TheRealFalcon commented on doc/rtd/explanation/format.rst:
```diff
@@ -3,18 +3,65 @@
-desired instance state. Cloud-config can be used to define how an instance
-should be configured in a human-friendly format.
+| **Header:** #cloud-config
+| **Content-Type:** text/cloud-config
```
Yeah...I was having a hard time figuring out how to fit these in. I think these are really both pieces of reference information, but I didn't think it made sense to have reference pages that only list header and content-types for user data formats.

The thing is these content types also matter for the cloud config archive and part handlers.

I get that these formats (along with the MIME archive) are going to much less used and basically "expert mode" for anybody using them...so...maybe I can try to cook up a reference page with enough info.

TheRealFalcon commented on doc/rtd/explanation/format.rst:
```diff
@@ -141,66 +225,128 @@ Create user data containing 3 shell scripts:
-using a MIME archive.
+.. note::
+    Use of Jinja templates is ONLY supported for cloud-config and user data
+    scripts. Jinja templates are not supported for cloud-boothooks or
```
Yep, good idea. https://github.com/canonical/cloud-init/issues/5554

TheRealFalcon commented on doc/rtd/explanation/format.rst:
```diff
@@ -141,66 +225,128 @@ Create user data containing 3 shell scripts:
+Gzip compressed content
+=======================
+
+| **Header** n/a
```
Yeah, this was never meant to call out generic file headers. It's for cloud-init required headers which is why it's n/a for gzip.

TheRealFalcon commented on doc/rtd/explanation/format.rst:
```diff
@@ -141,66 +225,128 @@ Create user data containing 3 shell scripts:
+=======================
+
+| **Header** n/a
+| **Content-Type** n/a
```
Yes, but given that it's a binary file, it can't be specified anywhere.

s-makin reviewed (approved):
+1 from me :)

aciba90 commented on doc/rtd/explanation/format.rst:
```diff
@@ -24,64 +69,113 @@ These things may include:
+
+* It is run very early in boot, even before the ``cc_bootcmd`` module
+* It is run on every boot
+* The environment variable ``INSTANCE_ID`` is set to the current instance ID
```
`INSTANCE_ID` is also provided for [cc_bootcmd](https://github.com/canonical/cloud-init/blob/779dd6b009f8a568fe74e98df9803ad590b83044/doc/module-docs/cc_bootcmd/data.yaml#L5), could you please remove this item?

aciba90 reviewed (commented):
Many thanks for this effort. It looks great!

TheRealFalcon commented:
Not all comments have been addressed, but I wanted to push what I had so far.

TheRealFalcon commented on doc/rtd/explanation/format.rst:
```diff
@@ -24,64 +71,101 @@ These things may include:
 
-.. code-block::
+A user data script is a single shell script to be executed once per instance.
+User data scripts are run relatively late in the boot process, after most
```
I do think that our stages go long enough that it makes sense to place these relative to a specific module. I don't think it's confusing info if the user doesn't need it. Let me know if you think it makes sense.

TheRealFalcon commented on doc/rtd/explanation/format.rst:
```diff
@@ -24,64 +71,101 @@ These things may include:
+A cloud boothook is similar to a :ref:`user data script<user_data_script>`
+in that it is a shell script run on boot. The boothook is different in that:
+
+* It is run very early in boot, even before the ``cc_bootcmd`` module
```
I do think that our stages go long enough that it makes sense to place these relative to a specific module. I don't think it's confusing info if the user doesn't need it. Let me know if you think it makes sense.

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -24,64 +71,121 @@ These things may include:
---------------
+A user data script is a single shell script to be executed once per instance.
+User data scripts are run relatively late in the boot process, during the
+'modules:final' stage as part of the "cc_scripts_user" module.
```
I know that `modules:final" text that is commonly seen in the log, but I think it would be easier to standardize on the more generic "final" stage, since the "modules" entrypoint is really just an implementation detail that the user shouldn't need to know about.

Also, could we link to the respective stages on the boot order page both here and in the cloud boothook section?

TheRealFalcon commented:
I believe the only fix left is to move the part handler stuff. Since that requires a rebase and merge, I figured it'd be best to push first.

TheRealFalcon commented:
Most recent commit adds the part handler split. Tbh, I don't love how it's split, but I do think it makes sense to keep information in both places. What do people think of having the example in both locations?

I believe with this commit I have addressed all of the comments.

TheRealFalcon commented on doc/rtd/spelling_word_list.txt:
```diff
@@ -211,6 +212,7 @@ scaleway
 seedurl
 serverurl
 setup-keymap
+shellscript
```
I don't love this as in most places this would be a legitimate misspelling, but I couldn't find any way to ignore a spelling for a single line.

holmanb replied:
Would not be legitimate? 

I think Chad submitted a PR in the last day or two that ignored spelling for a single line.

TheRealFalcon replied:
> Not sure that I understand, did you mean to say "... would not be a legitimate mispelling"?

No. I think that most of the time we want "shell script" and so I think adding this has the potential to silence legitimate misspellings.

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -24,93 +59,132 @@ These things may include:
-:ref:`jinja templates <instancedata-Using>`.
+A user data script is a single shell script to be executed once per instance.
+User data scripts are run relatively late in the boot process, during the
+cloud-init's final stage as part of the "cc_scripts_user" module. When run,
```
I would suggest a link to the cc_scripts_user module, however it's really more of an implementation detail and that module's documentation isn't very informative.

holmanb replied:
Could we link to the final stage on the boot stages page from here? That would be useful for anybody who isn't familiar with cloud-init's ordering.

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -24,93 +59,132 @@ These things may include:
-User data scripts can optionally render cloud instance metadata variables using
-:ref:`jinja templates <instancedata-Using>`.
+A user data script is a single shell script to be executed once per instance.
+User data scripts are run relatively late in the boot process, during the
```
"during the cloud-init's final stage" sounds funny

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -3,18 +3,53 @@
+    #cloud-config
+    password: password
+    chpasswd:
+    expire: False
```
indent

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -24,93 +59,132 @@ These things may include:
-   Cloud config data can also render cloud instance metadata variables using
-   :ref:`jinja templates <instancedata-Using>`.
+See the :ref:`yaml_examples` section for a set of commented examples of
+supported cloud config formats.
```
> set of commented examples of supported cloud config formats.

I know that you didn't write this in this PR, but I don't think that it makes sense to call these "cloud config formats". 

Also, this seems like an appropriate place to link to the modules reference page as well. Could we do that while making this change?

How about something like this:
```rst
For more information, see the cloud config 
:ref:`example configurations page <link_to_examples_page>`
 or the cloud config :ref:`modules reference page<link_to_modules_page>`.
```

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -24,93 +59,132 @@ These things may include:
 
-User data scripts can optionally render cloud instance metadata variables using
-:ref:`jinja templates <instancedata-Using>`.
+A user data script is a single shell script to be executed once per instance.
```
It doesn't have to be a shell script does it? Wouldn't anything that can run with a shebang work (lua, perl, make, tcl, guile, etc)? Maybe just drop the word "shell"?

TheRealFalcon replied:
> Wouldn't anything that can run with a shebang work 

Isn't that by definition a shell script then? Shell doesn't just mean `sh` or one of its descendants. Most of your examples there are listed as scripting languages [here](https://en.wikipedia.org/wiki/Scripting_language#Notable_languages)

TheRealFalcon replied:
(ignore my other comment here if you saw it)

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -24,93 +59,132 @@ These things may include:
-Kernel command line
-===================
+A cloud boothook is similar to a :ref:`user data script<user_data_script>`
+in that it is a shell script run on boot. When run,
```
same comment as above about "shell"

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -24,93 +59,132 @@ These things may include:
-This is typically useful because user data is limited to ~16384 [#]_ bytes.
+.. _user_data_formats-mime_archive:
 
 MIME multi-part archive
```
How would you feel about ordering the "formats that deal with other user data formats" in order of increasing complexity in the list above and also in the sections starting here?

If that seems reasonable, I would suggest:

Include file
Jinja template
cloud config archive
Mime multi-part archive
Part handler

TheRealFalcon replied:
Yes, but I want to put the mime one first as cloud config archive is framed as an alternative to the mime type.

holmanb commented on doc/rtd/explanation/format.rst:
```diff
@@ -24,93 +59,132 @@ These things may include:
 
-Gzip compressed content
-=======================
+* It is run very early in boot, during the 'init' stage, before any
```
Could we link to the init stage on the boot stages page from here? That would be useful for anybody who isn't familiar with cloud-init's ordering.

holmanb reviewed (commented):
@TheRealFalcon I think this is really close. I left a couple more comments.
